atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownConcludedTransactions.java

```diff
@@ -116,4 +127,19 @@ private void ensureRangesCached(RangeSet<Long> timestampRanges) {
                 SafeArg.of("numAttempts", MAX_ATTEMPTS));
         throw new SafeIllegalStateException("Unable to ensure ranges of known concluded transactions were cached.");
     }
+
+    @Value.Immutable
+    interface Cache {
+        @Value.Parameter
+        ImmutableRangeSet<Long> ranges();
+
+        @Value.Lazy
+        default long lastKnownConcludedTs() {
+            return ranges().asRanges().stream()
+                    .filter(Range::hasUpperBound)
+                    .map(Range::upperEndpoint)
+                    .max(Comparator.naturalOrder())
+                    .orElse(0L);
```

Comment:
I _think_ we only use closed intervals in the range set, but either way we want to either
- codify this requirement somewhere
- or be defensive to be open intervals (and so, instead of `Range::upperEndpoint`, we might want to say "if upper bound type is closed, upperEndpoint else upperEndpoint - 1").